flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/forwardgroup/StreamNodeForwardGroup.java

```diff
@@ -0,0 +1,197 @@
+    }
+
+    @Override
+    public boolean isParallelismDecided() {
+        return parallelism > 0;
+    }
+
+    @Override
+    public int getParallelism() {
+        checkState(isParallelismDecided());
+        return parallelism;
+    }
+
+    @Override
+    public boolean isMaxParallelismDecided() {
+        return maxParallelism > 0;
+    }
+
+    @Override
+    public int getMaxParallelism() {
+        checkState(isMaxParallelismDecided());
+        return maxParallelism;
+    }
+
+    @VisibleForTesting
+    public int size() {
+        return chainedStreamNodeGroupsByStartNode.values().stream().mapToInt(List::size).sum();
+    }
+
+    public Iterable<StreamNode> getStartNodes() {
+        return chainedStreamNodeGroupsByStartNode.keySet();
+    }
+
+    public Iterable<List<StreamNode>> getChainedStreamNodeGroups() {
+        return chainedStreamNodeGroupsByStartNode.values();
+    }
+
+    /**
+     * Responds to merge targetForwardGroup into this and update the parallelism information for
+     * stream nodes in merged forward group.
+     *
+     * @param targetForwardGroup The forward group to be merged.
+     * @return whether the merge was successful.
+     */
+    public boolean mergeForwardGroup(StreamNodeForwardGroup targetForwardGroup) {
+        checkNotNull(targetForwardGroup);
+
+        if (targetForwardGroup == this) {
+            return true;
+        }
+
+        if (!ForwardGroupComputeUtil.canTargetMergeIntoSourceForwardGroup(
+                this, targetForwardGroup)) {
+            return false;
+        }
+
+        this.chainedStreamNodeGroupsByStartNode.putAll(
+                targetForwardGroup.chainedStreamNodeGroupsByStartNode);
+
+        Set<Integer> configuredParallelisms = new HashSet<>();
+        if (this.isParallelismDecided()) {
+            configuredParallelisms.add(this.getParallelism());
+        }
+        if (targetForwardGroup.isParallelismDecided()) {
+            configuredParallelisms.add(targetForwardGroup.getParallelism());
+        }
+
+        checkState(configuredParallelisms.size() <= 1);
+
+        if (configuredParallelisms.size() == 1) {
+            this.parallelism = configuredParallelisms.iterator().next();
+        }
+
+        Set<Integer> configuredMaxParallelisms = new HashSet<>();
+        if (this.isMaxParallelismDecided()) {
+            configuredMaxParallelisms.add(this.getMaxParallelism());
+        }
+        if (targetForwardGroup.isMaxParallelismDecided()) {
+            configuredMaxParallelisms.add(targetForwardGroup.getMaxParallelism());
+        }
+
+        if (!configuredMaxParallelisms.isEmpty()) {
+            this.maxParallelism = Collections.min(configuredMaxParallelisms);
+            checkState(
+                    parallelism == ExecutionConfig.PARALLELISM_DEFAULT
+                            || maxParallelism >= parallelism);
+        }
+
+        if (this.isParallelismDecided() || this.isMaxParallelismDecided()) {
```

Comment:
> It's better to introduce two methods `updateNodeParallelism` and `updateNodeMaxParallelism` and invoke them only when any value really changes. Besides that, looks to me here the invocation of `setMaxParallelism(int)` will always set the `parallelismConfigured` of a `StreamNode` to true. Not sure if it is expected? cc @JunRuiLee

Do you mean `setParallelism(int)`? The invocation of `setMaxParallelism(int)` wouldn't change the `parallelismConfigured`. And it seems expected to me, based on the following code: https://github.com/apache/flink/blob/77e7ac076c35d229fab89b8b6ae15138b0e41a22/flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java#L908C49-L909C95

Reply:
Yes I meant `setParallelism(int)`.
In the code from the shared link, it can work because the parallelism will be unset for a StreamNode without user specified parallelism. (https://github.com/apache/flink/blob/77e7ac076c35d229fab89b8b6ae15138b0e41a22/flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java#L856)

We can do it here because the parallelism is also unset for any StreamNode without user specified parallelism, before doing such forward group computing?